rust/chains/hyperlane-cosmos/src/providers/grpc.rs

```diff
@@ -182,35 +204,51 @@ impl WasmProvider for WasmGrpcProvider {
         Ok(account)
     }
 
-    async fn generate_raw_tx<T>(&self, payload: T, gas_limit: Option<U256>) -> ChainResult<Vec<u8>>
+    async fn simulate_raw_tx<I>(
+        &self,
+        msgs: I,
+        gas_limit: Option<U256>,
+    ) -> ChainResult<SimulateResponse>
     where
-        T: Serialize + Send + Sync,
+        I: IntoIterator<Item = cosmrs::Any> + Send + Sync,
     {
-        let account_info = self.account_query(self.signer.address()).await?;
-        let contract_addr = self.get_contract_addr()?;
+        let mut client = TxServiceClient::connect(self.get_conn_url()?).await?;
 
-        let msg = MsgExecuteContract {
-            sender: contract_addr.clone(),
-            contract: contract_addr.clone(),
-            msg: serde_json::to_string(&payload)?.as_bytes().to_vec(),
-            funds: vec![],
-        };
+        let tx_bytes = self.generate_raw_tx(msgs, gas_limit).await?;
+        let sim_req = tonic::Request::new(SimulateRequest { tx: None, tx_bytes });
+        let mut sim_res = client.simulate(sim_req).await.unwrap().into_inner();
+
+        // apply gas adjustment
+        sim_res.gas_info.as_mut().map(|v| {
+            v.gas_used = (v.gas_used as f32 * DEFAULT_GAS_ADJUSTMENT) as u64;
+            v
+        });
```

Comment:
is the simulation imprecise?

Reply:
@yorhodes the gas results unit and real osmo unit are different. so have to change